ChopChop/Models/Recipe/Recipe/Recipe.swift

```diff
@@ -0,0 +1,210 @@
+        assert(checkRepresentation())
+        self.difficulty = difficulty
+        assert(checkRepresentation())
+    }
+
+    // step related functions
+    func addStep(content: String) throws {
+        assert(checkRepresentation())
+        steps.append(try RecipeStep(content: content))
+        assert(checkRepresentation())
+    }
+
+    func removeStep(_ removedStep: RecipeStep) {
+        assert(checkRepresentation())
+        guard (steps.contains { $0 == removedStep }) else {
+            return
+        }
+
+        steps.removeAll { $0 == removedStep }
+        assert(checkRepresentation())
+    }
+
+    func reorderStep(movedStep: RecipeStep, isSwapUp: Bool) throws {
+        assert(checkRepresentation())
+        guard let idx = (steps.firstIndex { $0 == movedStep }) else {
+            throw RecipeError.nonExistentStep
+        }
+
+        if isSwapUp {
+            guard idx > 0 else {
+                throw RecipeError.invalidReorderSteps
+            }
+            steps[idx] = steps[idx - 1]
+            steps[idx - 1] = movedStep
+        } else {
+            guard idx < steps.count - 1 else {
+                throw RecipeError.invalidReorderSteps
+            }
+
+            steps[idx] = steps[idx + 1]
+            steps[idx + 1] = movedStep
+        }
+
+        assert(checkRepresentation())
+    }
+
+    // ingredient related functions
+    func addIngredient(name: String, quantity: Quantity) throws {
+        assert(checkRepresentation())
+        if let existingIngredient = ingredients.first(where: { $0.name == name }) {
+            try existingIngredient.add(quantity)
+        } else {
+            let addedIngredient = try RecipeIngredient(name: name, quantity: quantity)
+            ingredients.append(addedIngredient)
+        }
+        assert(checkRepresentation())
+    }
+
+    func removeIngredient(_ removedIngredient: RecipeIngredient) throws {
+        assert(checkRepresentation())
+        guard (ingredients.contains { $0 == removedIngredient }) else {
+            throw RecipeError.nonExistentIngredient
+        }
+
+        ingredients.removeAll { $0 == removedIngredient }
+        assert(checkRepresentation())
+    }
+
+    func updateIngredient(oldIngredient: RecipeIngredient, name: String, quantity: Quantity) throws {
+        // note there is no effect on steps on updating ingredients
+        assert(checkRepresentation())
+        guard let _ = ingredients.first(where: { $0.name == oldIngredient.name }) else {
+            throw RecipeError.nonExistentIngredient
+        }
+        if oldIngredient.name == name {
+            oldIngredient.updateQuantity(quantity)
+        } else {
+            guard let existingIngredient = ingredients.first(where: { $0.name == name }) else {
+                try oldIngredient.rename(name)
+                oldIngredient.updateQuantity(quantity)
+                return
+            }
+            try existingIngredient.add(quantity)
```

Comment:
Do you need to remove oldIngredient also?

Reply:
oh good catch!